### ISSUE TITLE
Batching / Transformer-based Parallelization for Cross-Encoders

### ISSUE THREAD

**Issue by bmw-friedrich-mayr** [vespa-engine/vespa#18882]
I opened this feature-request on the basis of this [slack thread](https://vespatalk.slack.com/archives/C01QNBPPNT1/p1630055029072900) and @jobergum 's [comment](https://vespatalk.slack.com/archives/C01QNBPPNT1/p1630056538087000?thread_ts=1630055504.080000&cid=C01QNBPPNT1).

As mentioned by many Vespa users in slack, it would be awesome if there was an example for batching for cross-encoders. (Maybe even with 100 documents for reranking, as batching speeds up cross-encoders tremendously)

As you probably noticed cross-encoders are very slow in Vespa. I know cross-encoder are considerably slower than bi-encoders. But I think Vespa misses a crucial thing that speeds them up tremendously, batching. Transformers are able to process multiple inputs in parallel. From my point of view Vespa fetches only one sample to a cross-encoder, which is detrimental for performance.

Parallel calculation with Hugging Face:

```
cross_model = '<MODEL>'
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch

model = AutoModelForSequenceClassification.from_pretrained(cross_model)
tokenizer = AutoTokenizer.from_pretrained(cross_model)
model.eval()

query = "How many citizens has Berlin?"
docs = ["Berlin has a population of 3.5 million in 2016, up slightly from 3.4 million in 2014", ...]
features = tokenizer([query for doc in docs], docs, padding=True, truncation=True, return_tensors="pt", max_length=128)

with torch.no_grad():
    scores = model(**features).logits
    print(scores)
```

**Comment by jobergum** [comment]
@bmw-friedrich-mayr 

I've started on this work and an Reranking in container example will be added to the [ms marco ranking sample application](https://github.com/vespa-engine/sample-apps/tree/master/msmarco-ranking). ETA mid next week.

**Comment by bmw-friedrich-mayr** [comment]
That's great. Thank you @jobergum 🙂!

**Comment by jobergum** [comment]
Just an observation on batching input sequences and cpu serving (Vespa is currently cpu only)

Consider this from [Optimizing bert model for intel cpu cores](https://cloudblogs.microsoft.com/opensource/2021/03/01/optimizing-bert-model-for-intel-cpu-cores-using-onnx-runtime-default-execution-provider/)

![image](https://user-images.githubusercontent.com/20928528/132830214-66c98a49-d5bc-4f46-a10a-75c9acf8f9bc.png)

If you look at the number of sequences/second (Y axis) and compare the batch sizes per input sequence length there is not much change in sequences per second.  When comparing batch of sizes 1,2,4,8. For a cross-encoder inputing both the query and the passage sequence length 128 is the relevant part of the graph. 

>As you probably noticed cross-encoders are very slow in Vespa.

The MiniLM model we used for the MS Marco Passage ranking has 6 layers, using int8 weights 

With 12 threads per search with batch size 1 and 24 sequence inputs (rerank count) of 128 sequence length the Vespa backend is able to do 480 inferences per second.

**Comment by jobergum** [comment]
The re-ranking example is ready, plus many other improvements to the mentioned sample application. I've moved all transformer query encoders to the stateless layer and removed the query document type which was used to compute the query encoders. This avoids network round trip and also would make it easier to use auto-scaling features as compute is moved out of the content nodes. 

We need some work in Vespa to be able to tune onnx-rt intra threads per model before pushing this to master as one wants to use different number of intra threads per model for optimal performance, e.g the query encoders operate with a short sequence. For example the COLBert multi representation model always uses sequence length 32 due to the padding of mask tokens up to max length, while the dense single representation model is dynamic up to max length. For the latter using more than 1 intra thread is just waste without real benefit. The all to all interaction model with longer sequence length 128 can benefit from more intra threads. 

So in short we will need to add support for configuring intra threads per model. 

Preliminary benchmarking shows no improvement in the passages/s with batch > 1 over doing batch size 1, maybe even less which is similar results as in the above comment.

**Comment by bmw-friedrich-mayr** [comment]
Thanks @jobergum. I'm amazed by your fast support. Do I understand correctly that batching has little effect in your experiment? If this is the case, there has to be something wrong. Otherwise reranking in Vespa would have to be much faster.

Do you pass the full batch through the model at once?
```
cross_model = '<MODEL>'
from transformers import AutoTokenizer, AutoModelForSequenceClassification
import torch

model = AutoModelForSequenceClassification.from_pretrained(cross_model)
tokenizer = AutoTokenizer.from_pretrained(cross_model)
model.eval()

query = "How many citizens has Berlin?"
docs = ["Berlin has a population of 3.5 million in 2016, up slightly from 3.4 million in 2014", ...]
features = tokenizer([query for doc in docs], docs, padding=True, truncation=True, return_tensors="pt", max_length=128)

with torch.no_grad():
    scores = model(**features).logits
    print(scores)
```

**Comment by jobergum** [comment]
@bmw-friedrich-mayr, 

Yes. I've compared batch size = 24 with different number of intra threads (1,2,4,8,12,24) with batch size 1 with external threads (12) for the same sequence length. By external threads I mean how we evaluate ONNX models in the stateful content nodes where we use multiple threads to run inference but with intra threads set to 1. 

 See https://github.com/vespa-engine/vespa/issues/18882#issuecomment-916764298, also 
[this blog post](https://medium.com/microsoftazure/accelerate-your-nlp-pipelines-using-hugging-face-transformers-and-onnx-runtime-2443578f4333) report similar results. 

![image](https://user-images.githubusercontent.com/20928528/133036762-7b88aab6-d3e4-409a-b7a7-75645b92be35.png)

To get higher number of inferences per second  when increasing batch size from 1 to 4 the latency of batch size 4 needs to be less than 4x the latency of batch size 1, if that is the case you get higher number of inferences per second using batching.

### CPU:

Sequence length 128, Batch = 1, ONNX Runtime CPU 189.56 ms
Sequence length 128, Batch = 4, ONNX Runtime CPU 674.05 

= 3.55x 

Sequence length 256, Batch = 1, ONNX Runtime CPU 360.91 ms
Sequence length 256, Batch = 4, ONNX Runtime CPU 1485.55 ms 

= 4.1x

### GPU/FP32 
Sequence length 128, Batch = 1, ONNX Runtime GPU/FP32 4.30 ms 
Sequence length 128, Batch = 4, ONNX Runtime GPU/Fp32 11.28 ms

= 2.6x 

Sequence length 256, Batch = 1, ONNX Runtime GPU/FP32 7.37ms 
Sequence length 256, Batch = 4, ONNX Runtime GPU/Fp32 20.34 

= 2.75x

In the case with GPU, you get a much higher inferences per second as compared to CPU, while with CPU you don't have the same effect.

**Comment by jobergum** [comment]
https://github.com/vespa-engine/sample-apps/pull/705 introduces an example using batch re-ranking stage

**Comment by jobergum** [comment]
Mentioned PR merged, go check out https://github.com/vespa-engine/sample-apps/blob/master/msmarco-ranking/passage-ranking.md 

See the ReRankingSearcher. Let me know if you have any questions on this before I close this one out. Thank you
